### ISSUE TITLE
macOS icon cannot be provided as icns

### ISSUE THREAD

**Issue by mjtorn** [godotengine/godot#28930]
**Godot version:**

3.0.x

**Issue description:**

See also #23968

Am I reading the code right in that it scales a given PNG to different sizes?

This is problematic, because Apple's [Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/macos/icons-and-images/app-icon/) have things to say like

> **Simplify your icon at smaller sizes.** There are fewer pixels to draw as icon size decreases. In your smaller icons, remove unnecessary features and exaggerate primary features so they remain clear. Even when a high-resolution size matches the pixel dimensions of a standard size, you should still consider simplifying the smaller rendered image.

We tried using an _icon.icns_ file, generated by `png2icns` and got the default Godot icon everywhere, except oddly enough an _icon.png_ image, which would not be used on osx, in the bar.

**Steps to reproduce:**
Add a proper Mac icon when exporting and check it out in Finder.

**Comment by mjtorn** [comment]
It seems I found 3.1 code through the given issue, and it did not cherry-pick cleanly onto 3.0. Doesn't matter.

It does seem that the fundamental problem is this:

```
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 537)                if (file == "Contents/Resources/icon.icns") {
0b806ee0fc (Juan Linietsky        2014-02-09 22:10:30 -0300 538)                        //see if there is an icon
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 539)                        String iconpath;
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 540)                        if (p_preset->get("application/icon") != "")
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 541)                                iconpath = p_preset->get("application/icon");
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 542)                        else
25678b1876 (Juan Linietsky        2017-07-19 17:00:46 -0300 543)                                iconpath = ProjectSettings::get_singleton()->get("application/config/icon");
52466d57e9 (Rémi Verschelde       2018-08-24 09:35:07 +0200 544) 
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 545)                        if (iconpath != "") {
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 546)                                Ref<Image> icon;
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 547)                                icon.instance();
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 548)                                icon->load(iconpath);
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 549)                                if (!icon->empty()) {
5b44f092f9 (Bastiaan Olij         2017-06-21 19:11:38 +1000 550)                                        _make_icon(icon, data);
0b806ee0fc (Juan Linietsky        2014-02-09 22:10:30 -0300 551)                                }
0b806ee0fc (Juan Linietsky        2014-02-09 22:10:30 -0300 552)                        }
0b806ee0fc (Juan Linietsky        2014-02-09 22:10:30 -0300 553)                        //bleh?
0b806ee0fc (Juan Linietsky        2014-02-09 22:10:30 -0300 554)                }
```

I can't be arsed to check the `String` api now to see if `endswith` is correct, but shouldn't it be along the lines of

```
if (!iconpath.endswith("icon.icns") {
    Ref<Image> icon;
    // ...
} else {
    // Do some file reading so `data` has a size
    // so we output the proper file into the export
    // a few lines down
}
```

Then you can create HIG-compliant icns files using eg. `png2icns` (as long as your individual pngs are ok) and if you don't want to, you can supply a Godot-compatible image as the icon for automatic scaling.

Am I on the right track? What did I get wrong?

Thanks!

PS.
What's up with "bleh?" there? Is it significant?

**Comment by akien-mga** [comment]
> It seems I found 3.1 code through the given issue, and it did not cherry-pick cleanly onto 3.0. Doesn't matter.

Well it does matter, it's best to reuse the code which already works in the 3.1 and master branches instead of trying to redo that same fix independently :)

I cherry-picked #23989 in my fork's `3.0` branch: https://github.com/akien-mga/godot/tree/3.0
The only change needed was to remove the `.write` changes from 0e29f7974b59e4440cf02e1388fb9d8ab2b5c5fd.

I did not test that it works though, only that it builds fine. If you can confirm that it works, I can push it on `godotengine/3.0`.

**Comment by mjtorn** [comment]
Like I said, the fundamental problem seems to be that ICNS files can't be provided while exporting. Maybe I can try grabbing that patch of yours soon, but the PR in #23989 did not seem to do anything like what I proposed should be done...

So am I completely lost with how that should work or is there a facility already in place to introduce HIG-compliant ICNS icons for macOS?

Thanks!

**Comment by akien-mga** [comment]
I dont have any experience with macOS, maybe @bruvzg can clarify. The code in that PR does have conditional use of PNG, so I assume that the else block is ICNS, but I might be wrong.

**Comment by bruvzg** [comment]
> The code in that PR does have conditional use of PNG

#23989 is not directly related to this issue.

ICNS file is an archive with multiple icons of different sizes inside.

#23989 fixes generation of the small icons: big icons are stored as PNGs, small icons as RLE bitmaps, previously all sizes were PNGs.

All icon sizes in the ICNS file are generated from the same single source PNG file. Currently, there's no option to use used provided, pre-generated ICNS file and #23989 is not adding any way to do it.

> Am I on the right track? What did I get wrong?

Yes, checking icon source file extension with something like `iconpath.get_extension() == "icns"` and reading it to the `data` variable should work.

**Comment by mjtorn** [comment]
Thanks @bruvzg ! :) Really made my morning!

I had to do just a bit of backporting for 3.0. If you happen to be around while I'm still compiling, could you see if I made any mistages with the `icon->get_buffer(&data[0], icon->get_size())` thing?

https://github.com/mjtorn/godot/commit/93e0f1ef18ba35590392f8ae08cf2e90180d9153

**Comment by mjtorn** [comment]
I got it compiled, can verify the correct ICNS file is included in the .app, but when installing it to Applications, the minuscule logo looks like it's been scaled down from the big one. This may be Apple weirdness and not a bug in #28954 or my backport. Do you know about this @bruvzg ?

What does look like a bug, though, is when the game is run. The icon in the dock is our "canary" PNG, though I'm pretty sure the ICNS file should be used there as well...

I tried the [rm && killall](https://apple.stackexchange.com/questions/150576/dock-icon-not-refreshing-when-icns-file-is-changed-on-yosemite) solution, but it did not help.

Good job on the patch, but we're probably not quite done yet!

**Comment by bruvzg** [comment]
> but when installing it to Applications, the minuscule logo looks like it's been scaled down from the big one

It's probably Finder icon cache, try using `touch /Applications/YourAppName.app` command.

> What does look like a bug, though, is when the game is run. The icon in the dock is our "canary" PNG, though I'm pretty sure the ICNS file should be used there as well...

That's another issue. Dock icon is handled separately by [`OS.set_icon()`](https://github.com/godotengine/godot/blob/master/platform/osx/os_osx.mm#L1861-L1898) and project settings. AFAIK it is not capable of loading ICNS.

**Comment by mjtorn** [comment]
> It's probably Finder icon cache, try using touch /Applications/YourAppName.app command.

Sadly this did not help :( So strange...

> That's another issue. Dock icon is handled separately by OS.set_icon() and project settings. AFAIK it is not capable of loading ICNS.

A bug in Godot or a known macOS thing? We are obviously going to have a different PNG and using it will sweep the issue under the rug, but it'd be nice to know if that warrants opening a new issue to be fixed by a new PR...

**Comment by bruvzg** [comment]
> Sadly this did not help :( So strange...

Maybe issue with `.ICNS` file itself, Finder is quite picky for icons.
For better results try rendering all icon sizes manually and use `iconutil` dirctly instead of tools like `png2icns`.

1. Make a `NAME.iconset` folder with the following `.PNG` files:
	- `icon_16x16.png` and `icon_16x16@2x.png` (same design features as 16x16, but rendered at double resolution)
	- `icon_32x32.png` and `icon_32x32@2x.png`
	- `icon_128x128.png` and `icon_128x128@2x.png`
	- `icon_256x256.png` and `icon_256x256@2x.png`
	- `icon_512x512.png` and `icon_512x512@2x.png`
2. Use `/usr/bin/iconutil -c icns NAME.iconset` command to generate `.ICNS` file.

> A bug in Godot or a known macOS thing? We are obviously going to have a different PNG and using it will sweep the issue under the rug, but it'd be nice to know if that warrants opening a new issue to be fixed by a new PR...

Result of the platform-independent nature of Godot, I guess. There's same issue on Widows as well, you can specify `.ICO` file for the export, but window/taskbar icon is `.PNG` only.

**Comment by mjtorn** [comment]
As for the ICNS, our art director started looking for software he can use to keep all the loose ends in his own hands. I will look at `iconutil` after a lunch meeting and touch base with the AD to see if he found something, but what you said about `iconutil` looks a lot clearer than `png2icns`!

> Result of the platform-independent nature of Godot, I guess. There's same issue on Widows as well, you can specify .ICO file for the export, but window/taskbar icon is .PNG only.

Ok, I think we'll run with a 256x256 PNG, or some such size, on these and hope it scales fine.

Thanks! I'll let you know when I have something :)

**Comment by bruvzg** [comment]
As a workaround for the dock icon, you can try setting `Project Settings->Application->Config->Icon` to the empty string, this should force app to use default app icon (`.ICNS` specified at export).

**Comment by mjtorn** [comment]
@bruvzg thank you so much for the help!

My lunch meeting dragged on so I can't make changes to our export system, to accommodate that empty string, because I'm needed home to kick off the weekend. And 256x256 png isn't that bad as a worst-case scenario.

BUT I can confirm `iconutil` did the trick :) Looks amazing too :)

**Comment by bruvzg** [comment]
> I can't make changes to our export system, to accommodate that empty string, because I'm needed home to kick off the weekend.

Empty string doesn't work anyway, I was wrong it's resetting do default Godot logo in this case.

But I have drafted support for the Windows and macOS in my fork - https://github.com/bruvzg/godot/tree/native_icon_support (https://github.com/bruvzg/godot/commit/bca5973b92dd279ac86ce485e2ef54c1bca657fc).

It adds `Project Settings`->`Application`->`Config`->`macOS Native Icon` and `Windows Native Icon` settings and `OS.set_native_icon(filename)` API. (Windows part is untested and may be completely wrong, macOS part seems to work).

**Comment by mjtorn** [comment]
@bruvzg very cool! I backported it as https://github.com/mjtorn/godot/commit/bbf687a8dcb93d3e4f3176c0b5fdf0233f83edb3 and it compiles on Linux, macOS and Windows.

But sadly I found a bug in our game, which I would prefer to fix before testing actual exports again. It's a tough combination of `Node2D` and `Control` elements overlapping and confusing the tooltip, so I have no idea how long it'll take to fix.

I'll let you know about Windows findings as soon as possible!

Thanks!

**Comment by mjtorn** [comment]
@bruvzg sorry for not having more right now, but I got delayed also by forgetting to configure the `wine64` and `rcedit` paths.

With them in place, I got the errors `Cannot open icon file '/'` and `Fatal error: Unable to set icon`.

If this is something you recognize immediately, please let me know :)

This is too late in the day for further testing, but without Wine properly configured, the alt-tab and task bar icons are still incorrect - they look a bit like the task manager or something. Are they related to the executable icon in that way?

Thanks!

**Comment by mjtorn** [comment]
@bruvzg ok, I finally have something concrete. The mistakes with `wine64` and that were all mine, I had a typo in the path. The native icon is still broken, though.

As an attempt to fix it, I recreated the Windows icon file with the layers so that the largest is first, and added 24x24, which were missing before. This did not do anything for the native icon :(

Is there a Windows person who could take a look at it? Or do you have any clues as to what may be broken, knowing something apparently is?

**Comment by mjtorn** [comment]
Here's a screenshot

![2019-05-21 01 win-icon-broken](https://user-images.githubusercontent.com/78427/58074809-c8820180-7bae-11e9-8fdd-ddbb196df6fd.png)

**Comment by bruvzg** [comment]
Well, probably Windows code in https://github.com/bruvzg/godot/tree/native_icon_support is totally wrong, I do not have mush knowledge in WinAPI and unlikely will be able to help much with it.

**Comment by mjtorn** [comment]
Ok, thanks anyway :) I took this up on IRC, because I could imagine the offset/len calculations to be wrong for example. I'm not very experienced in this either, but it would be valuable to know if there's a way to validate the icon after getting, and the code could probably use some error checking like

```c++
} else {
    ERR_EXPLAIN("Offset for icon not found, error: " + format_error_message(GetLastError()));
    ERR_FAIL_V(ERR_INVALID_DATA);
}
```

It could also be that it's automatically validated, so `hicon` is unset, and that's what we're seeing in the screenshot, which would also benefit from an error.

I'll see what I can do...

**Comment by bruvzg** [comment]
Maybe it is problem with `LookupIconIdFromDirectoryEx` itself, see https://stackoverflow.com/a/41400807.

**Comment by mjtorn** [comment]
@bruvzg what a mess.. there's some resonable commenting there that maybe there isn't a bug, so that `ICONDIRENTRY` seems legit, but on the other hand your code should get the offsets correctly as well.

Then that link seems to back it up that you wouldn't have to know the exact byte count, as everyone's just passing in the maximum size, so that seems also legit.

I'll jump through the hoops of trying to detect errors from the code as it is and work from there.

PS.
https://blogs.msdn.microsoft.com/oldnewthing/20101022-00/?p=12473

`The image must be in 32bpp ARGB format (known to GDI+ as Pixel­Format­32bpp­ARGB).` Wonder if this is still true.

```
$ icotool -l ui/icons/icon.ico
--icon --index=1 --width=1024 --height=1024 --bit-depth=24 --palette-size=0
--icon --index=2 --width=512 --height=512 --bit-depth=24 --palette-size=0
--icon --index=3 --width=256 --height=256 --bit-depth=24 --palette-size=0
--icon --index=4 --width=128 --height=128 --bit-depth=24 --palette-size=0
--icon --index=5 --width=32 --height=32 --bit-depth=24 --palette-size=0
--icon --index=6 --width=24 --height=24 --bit-depth=24 --palette-size=0
--icon --index=7 --width=16 --height=16 --bit-depth=8 --palette-size=256
```

But it's made with `icotool` so I suppose it should work; I'm sure tons of people use `icotool` just fine.

**Comment by mjtorn** [comment]
@bruvzg using a borrowed laptop, I got some progress here. Would it make sense to have `*.ico` as a default export? Missing the icon in the export was failure number one.

Failure number two is "Could not create 16x16 icon, error: Error 0: The operation completed successfully" so thanks to Microsoft for being nice about `GetLastError()`... That error would be issued when `CreateIconFromResource(...)` fails, so yeah, something's definitely broken.

**Comment by bruvzg** [comment]
> Would it make sense to have *.ico as a default export? Missing the icon in the export was failure number one.

Yep, I missed this. Only "resources" (file extensions are registered by resource loaders) are exported by default, `*.ico` and `*.icns` should be added to the "export non-resorces files" filter (not sure if there's global filter somewhere or not).

**Comment by mjtorn** [comment]
That makes me wonder why it works so well on macOS, unless its export presets contain `*.icns` implicitly, which I'd doubt, but I also doubt it resolves the file from my file system path...

**Comment by mjtorn** [comment]
@bruvzg please see the screenshot below, messy as it may be. It works!

There are some things that need to happen on my part. I need to clean the code up a bit before pushing it. I also need to make an official build for our team to test. But generally I say we're close to done here.

How do you want to proceed? If I push a commit, would you cherry-pick it and make it into a PR with your commit so it can be included in Godot?

![isalade-icon-works](https://user-images.githubusercontent.com/78427/58175425-c00df180-7ca8-11e9-9d41-ea4cf08c3a1f.png)

**Comment by mjtorn** [comment]
@bruvzg Though this hasn't seen any formal QA yet, I think I should give you the commit for general review first and get back to you with what we find: https://github.com/mjtorn/godot/commit/cc7fe5341fb7ccb0a0fed8d3b23f8de986be870e

**Comment by bruvzg** [comment]
> Icons are so that the first is the largest, the last is the smallest

I have tested it with a dozen of random icons and this assumption seems to be wrong for most of them. I have added some extra checks, modified the code to search for largest and 16×16 icons with the highest colour count, replaced `CRASH_NOW` with `ERR_FAIL` and applied Godot's clang format code style rules. Here's PR: #29119

**Comment by mjtorn** [comment]
> I have tested it with a dozen of random icons and this assumption seems to be wrong for most of them.

Hah! Seems we have some kind of confirmation bias, maybe our AD's whatever Photoshop plugin (or something) just makes them that way ;) Or me trying things out with Gimp caused that.

> replaced CRASH_NOW with ERR_FAIL

Nice! I tried some of those macros and got complaints that a `void` function should not return, so I figured `CRASH_NOW` is the only way to go.

> applied Godot's clang format code style rules.

Ok, weird, cuz Visual Studio told me it found a clang format file and will apply it automatically. I got nothing, but good thing it got sorted.

------

The PR looks really sweet. Preliminary results indicate we have code that works, and I got crazy busy with other things now that we're out of the woods. Still I'll make a point of backporting your specific code to our 3.0 codebase and verify nothing breaks. If something breaks, it's most likely due to a fscked-up icon of our making, anyway.
